src/test/utils/TypeConverter.sol

```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+library TypeConverter {
+    function toString(address x) internal pure returns (string memory) {
+        bytes memory s = new bytes(40);
+        for (uint256 i = 0; i < 20; i++) {
+            bytes1 b = bytes1(uint8(uint256(uint160(x)) / (2**(8 * (19 - i)))));
+            bytes1 hi = bytes1(uint8(b) / 16);
+            bytes1 lo = bytes1(uint8(b) - 16 * uint8(hi));
```

Comment:
This is where the modulus/remainder operator comes in handy:

`bytes1 lo = bytes1(uint8(b) % 16);`

Since the 16 is a power of two, both of these can also be done with binary operations

```
bytes1 hi = bytes1(uint8(b) >> 4);
bytes1 lo = bytes1(uint8(b) & 0xf);
```

I'm just mentioning this out of interest, no need to gas-golf this function.